File: apps/windows/TeslaSync.App/shared-surfaces/DataTableResizer.Model.cs
using System.Globalization;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Canonical metadata + the pure clamp/label adapter for the column-resize handle — the native mirror of the web
/// <c>DataTableResizer</c> (web/src/components/ui/DataTableResizer.tsx). The web component is a fully presentational
/// drag handle on the right edge of a resizable <c>&lt;th&gt;</c>: it follows the WAI-ARIA Authoring Practices
/// "Window Splitter" pattern (<c>role="separator"</c> + <c>aria-orientation="vertical"</c> +
/// <c>aria-valuenow/min/max</c> + <c>tabIndex=0</c>), reports a continuous width while the pointer drags
/// (<c>onResize</c>) and a final width on release (<c>onResizeEnd</c>), and is keyboard-operable
/// (Left/Right nudge ±8px, Home resets to 80px, End maxes out). It reads no network data and renders no titles of
/// its own, so this carries the diagnostics slug, the automation id, the ARIA role/orientation contract, the
/// default bound + keyboard constants (web prop defaults), the single i18n key behind the accessible label (web
/// <c>label ?? `Resize column ${columnKey}`</c>) and the clamp maths the drag / keyboard / automation paths share.
/// UI-free so every value is asserted headlessly.
/// </summary>
public static class DataTableResizerRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "DataTableResizer";

    /// <summary>
    /// The root automation id the view stamps on itself. The web component declares no <c>data-testid</c> (it is an
    /// anonymous splitter handle), so this is the native-only stable hook UI-automation tests target.
    /// </summary>
    public const string RootAutomationId = "data-table-resizer";

    /// <summary>
    /// ARIA role the surface reproduces — the WAI-ARIA Window Splitter (web <c>role="separator"</c>). The native peer
    /// reports <c>AutomationControlType.Separator</c> and a RangeValue pattern to mirror the value semantics.
    /// </summary>
    public const string SeparatorRole = "separator";

    /// <summary>ARIA orientation the splitter declares — vertical (web <c>aria-orientation="vertical"</c>).</summary>
    public const string Orientation = "vertical";

    /// <summary>Default minimum column width in pixels when resizing (web <c>minWidth = 60</c>).</summary>
    public const int DefaultMinWidth = 60;

    /// <summary>Default maximum column width in pixels when resizing (web <c>maxWidth = 800</c>).</summary>
    public const int DefaultMaxWidth = 800;

    /// <summary>Keyboard nudge increment in pixels (web Left/Right shrink/grow by 8px).</summary>
    public const int KeyboardStep = 8;

    /// <summary>The width Home resets the column to, in pixels (web <c>Home</c> → 80px).</summary>
    public const int HomeWidth = 80;

    /// <summary>
    /// i18n key behind the accessible label (web <c>label ?? `Resize column ${columnKey}`</c>). The catalog stores the
    /// value with the i18next token <c>{{col}}</c>; the .NET fallback below uses the positional token <c>{0}</c>, so the
    /// resolved string is composed with <see cref="string.Format(IFormatProvider, string, object?)"/>.
    /// </summary>
    public const string ResizeLabelKey = "translation.table.columns.resizeLabel";

    /// <summary>
    /// English fallback for <see cref="ResizeLabelKey"/> in .NET positional form — the transform of the web
    /// <c>`Resize column ${columnKey}`</c> template into <c>"Resize column {0}"</c>, matching the P1/S10 catalog entry
    /// <c>table.columns.resizeLabel</c> ("Resize column {{col}}").
    /// </summary>
    public const string ResizeLabelFallback = "Resize column {0}";

    /// <summary>
    /// The accessible name for the handle — the native port of the web
    /// <c>aria-label={label ?? `Resize column ${columnKey}`}</c>. A non-blank <paramref name="labelOverride"/> is used
    /// verbatim (trimmed); otherwise the column key is composed into the localized "Resize column {col}" template.
    /// </summary>
    /// <param name="columnKey">The column key the handle resizes (web <c>columnKey</c>).</param>
    /// <param name="labelOverride">An explicit accessible label, or null/blank to compose from the column key (web <c>label</c>).</param>
    /// <param name="localizer">The i18n facade the default label resolves through.</param>
    public static string ResolveAccessibleName(string? columnKey, string? labelOverride, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        if (!string.IsNullOrWhiteSpace(labelOverride))
        {
            return labelOverride.Trim();
        }

        string template = localizer.GetString(ResizeLabelKey, ResizeLabelFallback);
        string key = (columnKey ?? string.Empty).Trim();
        return string.Format(CultureInfo.CurrentCulture, template, key).Trim();
    }
}

/// <summary>
/// The pure clamp helpers behind the resize handle — the native port of the web component's inline
/// <c>clamp</c> (web/src/components/ui/DataTableResizer.tsx L43-L46:
/// <c>Math.max(minWidth, Math.min(maxWidth, Math.round(n)))</c>). Every method is static, culture-independent and
/// side-effect-free so the handle's behaviour is unit-tested without a view or a UI thread. Rounding mirrors
/// JavaScript's <c>Math.round</c> for the non-negative pixel widths the surface produces.
/// </summary>
public static class DataTableResizerMath
{
    /// <summary>
    /// Clamp a raw width to the integer range [<paramref name="minWidth"/>, <paramref name="maxWidth"/>], rounding to
    /// the nearest pixel first (web <c>Math.max(minWidth, Math.min(maxWidth, Math.round(n)))</c>). When the bounds are
    /// inverted the minimum wins, exactly like the chained JS min/max.
    /// </summary>
    /// <param name="value">The raw width in pixels (typically <c>startWidth + dragDelta</c>).</param>
    /// <param name="minWidth">The minimum allowed width.</param>
    /// <param name="maxWidth">The maximum allowed width.</param>
    public static int Clamp(double value, int minWidth, int maxWidth)
    {
        long rounded = (long)Math.Round(value, MidpointRounding.AwayFromZero);
        long bounded = Math.Min(maxWidth, rounded);
        return (int)Math.Max(minWidth, bounded);
    }
}

/// <summary>
/// Pure projection of the resize handle's render inputs — the native port of the web component body
/// (web/src/components/ui/DataTableResizer.tsx). It clamps the current <see cref="Width"/> into
/// [<see cref="MinWidth"/>, <see cref="MaxWidth"/>] (so <c>aria-valuenow</c> is always in range), carries the
/// <see cref="IsDragging"/> flag that drives the handle's highlight (web <c>dragging &amp;&amp; 'opacity-100 …'</c>)
/// and resolves the <see cref="AccessibleName"/> (web <c>aria-label</c>). Kept a value type so the adapter is
/// unit-testable and snapshot-comparable without a view-model or a UI thread.
/// </summary>
public readonly record struct DataTableResizerProjection
{
    private DataTableResizerProjection(
        string columnKey,
        int width,
        int minWidth,
        int maxWidth,
        bool isDragging,
        string accessibleName)
    {
        ColumnKey = columnKey;
        Width = width;
        MinWidth = minWidth;
        MaxWidth = maxWidth;
        IsDragging = isDragging;
        AccessibleName = accessibleName;
    }

    /// <summary>The column key the handle resizes (web <c>columnKey</c>).</summary>
    public string ColumnKey { get; }

    /// <summary>The current column width in pixels, clamped to the bounds (web <c>aria-valuenow</c>).</summary>
    public int Width { get; }

    /// <summary>The minimum allowed width in pixels (web <c>aria-valuemin</c> / <c>minWidth</c>).</summary>
    public int MinWidth { get; }

    /// <summary>The maximum allowed width in pixels (web <c>aria-valuemax</c> / <c>maxWidth</c>).</summary>
    public int MaxWidth { get; }

    /// <summary>Whether a drag is in progress, driving the handle's highlight (web <c>dragging</c>).</summary>
    public bool IsDragging { get; }

    /// <summary>The accessible name the splitter announces (web <c>aria-label</c>).</summary>
    public string AccessibleName { get; }

    /// <summary>
    /// Project the render inputs, reproducing the web component body: the width is rounded + clamped into the bounds,
    /// and the accessible name resolves the label override or the localized "Resize column {col}" default.
    /// </summary>
    /// <param name="columnKey">The column key the handle resizes (web <c>columnKey</c>).</param>
    /// <param name="width">The current (raw) column width in pixels (web <c>width</c>).</param>
    /// <param name="minWidth">The minimum allowed width (web <c>minWidth</c>).</param>
    /// <param name="maxWidth">The maximum allowed width (web <c>maxWidth</c>).</param>
    /// <param name="labelOverride">An explicit accessible label, or null/blank to compose from the column key (web <c>label</c>).</param>
    /// <param name="isDragging">Whether a drag is in progress (web <c>dragging</c>).</param>
    /// <param name="localizer">The i18n facade the default label resolves through.</param>
    public static DataTableResizerProjection Project(
        string? columnKey,
        double width,
        int minWidth,
        int maxWidth,
        string? labelOverride,
        bool isDragging,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        string key = (columnKey ?? string.Empty).Trim();
        return new DataTableResizerProjection(
            key,
            DataTableResizerMath.Clamp(width, minWidth, maxWidth),
            minWidth,
            maxWidth,
            isDragging,
            DataTableResizerRegistration.ResolveAccessibleName(key, labelOverride, localizer));
    }
}

/// <summary>
/// The sink the handle announces width changes through — the native seam for the web <c>onResize</c> (continuous,
/// every pointer move / keyboard step) and <c>onResizeEnd</c> (committed, on pointer release / per keyboard step).
/// Hosts wire it to their column-width state + persistence; tests record the calls. Pass
/// <see cref="NoOpColumnResizeSink.Instance"/> when nothing is wired (the web no-op callback equivalent).
/// </summary>
public interface IColumnResizeSink
{
    /// <summary>A continuous width update while resizing (web <c>onResize(next)</c>).</summary>
    /// <param name="columnKey">The column key being resized.</param>
    /// <param name="width">The new clamped width in pixels.</param>
    void OnResize(string columnKey, int width);

    /// <summary>The committed final width (web <c>onResizeEnd(final)</c>); use for persistence.</summary>
    /// <param name="columnKey">The column key that was resized.</param>
    /// <param name="width">The committed clamped width in pixels.</param>
    void OnResizeEnd(string columnKey, int width);
}

/// <summary>
/// An <see cref="IColumnResizeSink"/> that ignores every change — the native analogue of mounting the web component
/// with a no-op <c>onResize</c> and no <c>onResizeEnd</c>. Lets a handle be hosted in a gallery / design host without
/// wiring persistence.
/// </summary>
public sealed class NoOpColumnResizeSink : IColumnResizeSink
{
    /// <summary>The shared singleton instance.</summary>
    public static NoOpColumnResizeSink Instance { get; } = new();

    private NoOpColumnResizeSink()
    {
    }

    /// <inheritdoc />
    public void OnResize(string columnKey, int width)
    {
        // Intentionally ignored: no host is wired to the continuous width.
    }

    /// <inheritdoc />
    public void OnResizeEnd(string columnKey, int width)
    {
        // Intentionally ignored: no host is wired to the committed width.
    }
}

/// <summary>
/// PII-safe diagnostics for the resize handle (P1/S11 diagnostics contract). The surface carries only a transient
/// column width and a column key, so the collector records nothing but the operational <c>view.opened</c> event with
/// the surface slug — never a width or column key. Thread-safe; mirrors the peer surfaces' collectors.
/// </summary>
public sealed class DataTableResizerDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public DataTableResizerDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=DataTableResizer</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={DataTableResizerRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/DataTableResizer/DataTableResizerModel.kt
// Pure, framework-free model + bounds arithmetic + label resolver + diagnostics for the DataTableResizer shared
// surface — the native analogue of every decision the web component makes (web/src/components/ui/DataTableResizer.tsx)
// before Compose paints the handle. No Compose, no Android, no HTTP: every declaration here is unit-tested
// off-device in the :android:testReleaseUnitTest gate, keeping the composable a thin render layer over it.
//
// What the web source is (and therefore the COMPLETE behaviour this surface reproduces): a drag handle that
// resizes a `<th>`. Its only logic is the clamp `Math.max(minWidth, Math.min(maxWidth, Math.round(n)))` applied
// to (a) a pointer drag (`startWidth + (clientX - startX)`, emitted continuously through `onResize`, the final
// value persisted on release through `onResizeEnd`) and (b) the WAI-ARIA "Window Splitter" keyboard map
// (ArrowLeft −8, ArrowRight +8, Home → 80, End → maxWidth, each emitting `onResize` AND `onResizeEnd`). It also
// exposes `aria-valuenow/min/max` and an `aria-label` that defaults to `Resize column ${columnKey}`. It fetches
// nothing and owns no data.
//
// Why the generic data-surface states (loading / empty / error / stale / offline) are intentionally absent: this
// surface performs no query — it is a controlled affordance whose width the parent table owns. There is nothing
// to be loading, to be empty, to fail, to go stale, or to be offline, so inventing those states would model an
// async dependency the web spec does not have (honesty covenant: no scope narrowing, no silent drift). The
// surface's REAL, fully-reproduced states are the handle's interaction states — idle, keyboard-focused, and
// actively dragging — crossed with the value reaching its min / max bound. Each is reduced here in [ResizeBounds]
// and asserted off-device, doubling as the per-state snapshot. The presentational precedents are the sibling
// Checkbox / StaggerContainer / Accordion / TimelineScrubber surfaces (composable + pure model, no Source/ViewModel).
//
// SI boundary (unit-conversion instructions): the web works in CSS pixels; the native analogue is the
// density-independent pixel (dp), so the canonical width is modelled as an Int dp count and the composable
// converts pointer deltas with the screen density. There is no physical-unit conversion (km/mi, °C/°F, …) here —
// a column width is a layout measurement, not a vehicle datum — so the surface touches no SI boundary.
//
// i18n (P1/S10): the web owns one string, the `aria-label` default `Resize column ${columnKey}`. It resolves
// here by-name through the i18n facade ([resolveOptional], the native mirror of i18next `t(key, default)`) with
// the English [DataTableResizerDefaults.RESIZE_COLUMN_TEMPLATE] fallback, exactly like the sibling Accordion
// surface's native-only affordances — so there is no hardcoded English literal in the render path and a catalog
// translation overrides it without a code change.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/DataTableResizer — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName` is
// suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.datatableresizer

import io.teslasync.shared.core.diagnostics.Logger
import kotlin.math.roundToInt

/**
 * Diagnostics surface slug emitted with the `view.opened` event (P1/S11). Carries no column key and no width —
 * only this constant identifier — so a diagnostics line can never leak which column the user is resizing.
 */
const val DATA_TABLE_RESIZER_SLUG: String = "DataTableResizer"

/** Default minimum column width in dp — the native mirror of the web `minWidth = 60`. */
const val DEFAULT_MIN_WIDTH_DP: Int = 60

/** Default maximum column width in dp — the native mirror of the web `maxWidth = 800`. */
const val DEFAULT_MAX_WIDTH_DP: Int = 800

/** Keyboard nudge increment in dp — the web ArrowLeft/ArrowRight `± 8`. */
const val KEYBOARD_STEP_DP: Int = 8

/** The width Home snaps to in dp — the web `Home` key `clamp(80)`. */
const val HOME_WIDTH_DP: Int = 80

/** Resource name for the resize-handle accessible label (by-name; absent ⇒ the English template fallback). */
const val KEY_RESIZE_COLUMN: String = "translation_dataTableResizer_resizeColumn"

/**
 * Canonical registry metadata for the DataTableResizer surface. The diagnostics [SLUG] is emitted with the
 * one-shot `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`DataTableResizer`).
 */
object DataTableResizerRegistration {
    /** Stable surface id (kebab-case), also the test tag the composable stamps on its handle. */
    const val ID: String = "data-table-resizer"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = DATA_TABLE_RESIZER_SLUG
}

/**
 * Surface defaults that are not themselves Compose values — the web prop defaults plus the English label
 * template. The Dp-typed `minWidth` / `maxWidth` defaults live in the composable (a Dp needs Compose); these are
 * the framework-free numbers the composable derives them from, so the bound arithmetic stays unit-testable.
 */
object DataTableResizerDefaults {
    /** Default minimum width in dp (web `minWidth = 60`). */
    const val MIN_WIDTH_DP: Int = DEFAULT_MIN_WIDTH_DP

    /** Default maximum width in dp (web `maxWidth = 800`). */
    const val MAX_WIDTH_DP: Int = DEFAULT_MAX_WIDTH_DP

    /** Keyboard step in dp (web `± 8`). */
    const val STEP_DP: Int = KEYBOARD_STEP_DP

    /** Home-key width in dp (web `clamp(80)`). */
    const val HOME_DP: Int = HOME_WIDTH_DP

    /**
     * English fallback for the accessible label — a positional format template (`%1$s` ⇒ the column key) that is
     * the native mirror of the web default `Resize column ${columnKey}`. Used only when the catalog key
     * [KEY_RESIZE_COLUMN] is absent (the i18next `t(key, default)` default argument), so it is never a hardcoded
     * literal in the render path.
     */
    const val RESIZE_COLUMN_TEMPLATE: String = "Resize column %1\$s"
}

/**
 * One of the WAI-ARIA "Window Splitter" keyboard commands the handle accepts — the native mirror of the web
 * `onKeyDown` switch. [Shrink] / [Grow] nudge by [KEYBOARD_STEP_DP]; [Home] snaps to [HOME_WIDTH_DP]; [End]
 * maxes out at the bound's maximum. Pure (no Compose [androidx.compose.ui.input.key.Key]) so the command → width
 * resolution is unit-tested off-device; the composable maps physical keys onto these cases.
 */
enum class ResizeCommand {
    /** ArrowLeft — shrink one keyboard step. */
    Shrink,

    /** ArrowRight — grow one keyboard step. */
    Grow,

    /** Home — snap to the home width. */
    Home,

    /** End — grow to the maximum width. */
    End,
}

/**
 * The inclusive width bounds for a column, in dp — the native mirror of the web `minWidth` / `maxWidth` props.
 * All resolution flows through [clamp] so the rounded, clamped result matches the web
 * `Math.max(minWidth, Math.min(maxWidth, Math.round(n)))` exactly. Degenerate bounds (a max below the min, which
 * the web never guards) are normalised through [effectiveMax] so the model never throws.
 *
 * @property minWidthDp the smallest allowed width in dp (web `minWidth`).
 * @property maxWidthDp the largest allowed width in dp (web `maxWidth`).
 */
data class ResizeBounds(
    val minWidthDp: Int = DEFAULT_MIN_WIDTH_DP,
    val maxWidthDp: Int = DEFAULT_MAX_WIDTH_DP,
) {
    /** The max raised to at least the min, so [clamp] never coerces into an empty range. */
    val effectiveMax: Int
        get() = if (maxWidthDp < minWidthDp) minWidthDp else maxWidthDp

    /**
     * Round [valueDp] to the nearest whole dp and clamp it into `[minWidthDp, effectiveMax]` — the faithful port
     * of the web `clamp = Math.max(minWidth, Math.min(maxWidth, Math.round(n)))`.
     */
    fun clamp(valueDp: Float): Int = valueDp.roundToInt().coerceIn(minWidthDp, effectiveMax)

    /** Clamp an already-integer dp width (the keyboard/drag commit path). */
    fun clamp(valueDp: Int): Int = valueDp.coerceIn(minWidthDp, effectiveMax)

    /** [currentDp] grown/shrunk by [deltaDp], clamped — backs the keyboard step (web `clamp(width ± 8)`). */
    fun nudge(
        currentDp: Int,
        deltaDp: Int,
    ): Int = clamp(currentDp + deltaDp)

    /** The Home-key width — the web `clamp(80)`. */
    fun homeWidth(): Int = clamp(HOME_WIDTH_DP)

    /** The End-key width — the web `clamp(maxWidth)`, i.e. the (normalised) maximum. */
    fun endWidth(): Int = effectiveMax

    /**
     * Resolve the new width for a keyboard [command] applied to [currentDp] — the native mirror of the web
     * `onKeyDown` switch (ArrowLeft/ArrowRight/Home/End), each already clamped.
     */
    fun applyCommand(
        currentDp: Int,
        command: ResizeCommand,
    ): Int =
        when (command) {
            ResizeCommand.Shrink -> nudge(currentDp, -KEYBOARD_STEP_DP)
            ResizeCommand.Grow -> nudge(currentDp, KEYBOARD_STEP_DP)
            ResizeCommand.Home -> homeWidth()
            ResizeCommand.End -> endWidth()
        }
}

/**
 * Substitute [columnKey] into a resolved label [template] — the native mirror of the web
 * `Resize column ${columnKey}` interpolation. The template is a positional format string (`%1$s`), so a catalog
 * translation can reorder the column name for its grammar. Pure so it is unit-tested off-device.
 */
fun resizeColumnLabel(
    template: String,
    columnKey: String,
): String = template.format(columnKey)

/**
 * The accessible label the handle announces — the native mirror of the web `label ?? \`Resize column
 * ${columnKey}\``. A non-blank caller [override] (web `label`) wins; otherwise [columnKey] is formatted into the
 * resolved [template]. Pure so the precedence is unit-tested off-device.
 */
fun resolvedResizeLabel(
    override: String?,
    template: String,
    columnKey: String,
): String = override?.takeIf { it.isNotBlank() } ?: resizeColumnLabel(template, columnKey)

/**
 * Reproduces i18next's `t(key, default)` against the native i18n facade: returns the [lookup] result for
 * [resourceName] when it resolves to a non-blank string, otherwise the English [fallback]. [lookup] is a thin
 * seam over the Android string catalog in production (an optional by-name resource read) and a map in tests, so
 * the resolve-or-fallback decision stays pure and unit-tested.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the column
 * key, the width, or the bounds — so a diagnostics line can never leak which column the user is resizing or how
 * wide they made it. Kept free of Compose so it is unit-tested with a recording [Logger].
 */
object DataTableResizerDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = DATA_TABLE_RESIZER_SLUG

    /** The one-shot event emitted once when the surface opens. */
    const val EVENT_VIEW_OPENED: String = "view.opened"

    /** The structured-field key carrying the surface slug on the diagnostic. */
    const val FIELD_SURFACE: String = "surface"

    /**
     * Emits the one PII-safe `view.opened` diagnostic carrying only the surface [SLUG]. Call from the
     * composable's first-composition effect.
     */
    fun recordViewOpened(logger: Logger) {
        logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SURFACE to SLUG))
    }
}
